File: java/0309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {

    public int maxProfit(int[] prices) {
        Map<String, Integer> cache = new HashMap<>();
        return dfs(prices, cache, 0, true);
    }

    public int dfs(
        int[] prices,
        Map<String, Integer> cache,
        int index,
        boolean buying
    ) {
        if (index >= prices.length) {
            return 0;
        }
        String key = index + "-" + buying;

        if (cache.containsKey(key)) {
            return cache.get(key);
        }

        int cooldown = dfs(prices, cache, index + 1, buying);
        int buyOsell = Integer.MIN_VALUE;

        if (buying) {
            buyOsell = dfs(prices, cache, index + 1, !buying) - prices[index];
        } else {
            buyOsell = dfs(prices, cache, index + 2, !buying) + prices[index];
        }

        cache.put(key, Math.max(buyOsell, cooldown));
        return cache.get(key);
    }
}

File: rust/0309-best-time-to-buy-and-sell-stock-with-cooldown.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut p1 = 0;
        let mut p2 = 0;

        for x in 1..=prices.len()-1 {
            let temp = p1;
            p1 = std::cmp::max(p1+prices[x]-prices[x-1], p2);
            p2 = std::cmp::max(temp, p2);
        }

        return std::cmp::max(p1, p2)
    }
}
